Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp

```diff
@@ -671,6 +681,61 @@ void TreeBuilder::generate_missing_parents(NodeWithStyle& root)
         else
             parent.append_child(*wrapper);
     }
+
+    return table_roots_to_wrap;
+}
+
+template<typename Matcher, typename Callback>
+static void for_each_child_box_matching(Box& parent, Matcher matcher, Callback callback)
+{
+    parent.for_each_child_of_type<Box>([&](Box& child_box) {
+        if (matcher(child_box))
+            callback(child_box);
+    });
+}
+
+static void fixup_row(Box& row_box, HashMap<TableFormattingContext::GridPosition, bool> const& grid_occupancy, size_t row_index, size_t column_count)
+{
+    bool missing_cells_run_has_started = false;
+    for (size_t column_index = 0; column_index < column_count; ++column_index) {
+        if (grid_occupancy.contains({ column_index, row_index })) {
+            VERIFY(!missing_cells_run_has_started);
+            continue;
+        }
+        missing_cells_run_has_started = true;
+        auto cell_element = row_box.heap().template allocate<HTML::HTMLTableCellElement>(
+            row_box.document().realm(),
+            row_box.document(),
+            DOM::QualifiedName { HTML::TagNames::td, ""sv, Namespace::HTML });
+        cell_element->set_is_generated();
+        MUST(row_box.dom_node()->append_child(cell_element)); // FIXME propagate errors
+        [[maybe_unused]] auto invalidation_info = cell_element->recompute_style();
+        auto cell_box = row_box.heap().template allocate_without_realm<BlockContainer>(row_box.document(), cell_element, *cell_element->computed_css_values());
+        row_box.append_child(cell_box);
+    }
 }
 
+void TreeBuilder::missing_cells_fixup(Vector<JS::Handle<Box>> const& table_root_boxes)
+{
+    // Implements https://www.w3.org/TR/css-tables-3/#missing-cells-fixup.
+    for (auto& table_box : table_root_boxes) {
+        Layout::LayoutState throwaway_state;
+        Layout::TableFormattingContext throwaway_context(throwaway_state, *table_box, nullptr);
```

Comment:
let's move `TableFormattingContext::calculate_row_column_grid` into a function that could be used outside of TFC instead of doing layout during tree building.

Reply:
That function is also building `m_cells`, `m_columns` and `m_rows` besides building the occupancy grid, which makes it rather challenging to separate it without repeating code. How would you go about it?

Also I don't think it's fair to describe what we're doing as doing layout. We're not calling `TableFormattingContext::run` at any point during tree building.

Reply:
> That function is also building m_cells, m_columns and m_rows besides building the occupancy grid, which makes it rather challenging to separate it without repeating code. How would you go about it?

I don't expect to see any repeating code because basically you just need to make `calculate_row_column_grid` to be a separate function instead of TFC class method and pass `m_cells`, `m_columns` and `m_rows` as parameters. does that make sense?

Reply:
I disagree that'd be an improvement, but I'll apply it for the sake of moving this PR forward. Where would you put that separate function?

Reply:
let's create a new file called `TableGrid.h` and put it in there.

Reply:
Done.